pkg/plugins/tsdb.go

```diff
@@ -358,3 +358,12 @@ func convertDataTimePoint(point DataTimePoint) (t *time.Time, f *float64) {
 	}
 	return
 }
+
+type ErrTooManyRequests struct {
```

Comment:
I created this as a struct instead of just static error, because if it was a static error like `var TooManyRequests = errors.new("Too many requests")` it won't be possible to check it if it was wrapped by `%w`. Also I still want to provide some detail error message (the `Details` property) because otherwise it is not clear if this is error from Grafana or not and I think we should be explicit that this is an AWS limit and it has to be fixed there.

Reply:
Make sense :+1:

Reply:
Trying to understand things better. Do you want to provide a message to the user with just the raw error coming from AWS/CloudWatch or do you actually want to provide a custom message to the user explaining what happened, why and how to resolve things?

Or would the error just trigger a retry and the end user would never see the message?

Reply:
so first we retry until some timeout or number of retries, but after that the error should be shown why in the end we did not succeed with the query. At that point we could show just `Too many requests` message but that seems confusing. User run just single query and my first thought would be some limit in Grafana itself. So it this case I would like to show something like `Cloud Watch Logs max concurrent queries limit [10] exceeded` or something like that. So I thought about altering the original raw error a bit to something more user friendly and put that into Error.Details. But main point is to make clear this originated in Cloud Watch logs not in Grafana.

Reply:
I'm starting to doubt on my suggestion that 429 HTTP status makes sense to be returned when this happens since as you state it's not really a limit  enforced by Grafana, it's coming from CloudWatch. 

Would a 400 status work for all "plugin errors", but with a payload that could include more details in the response such like a `code`/`key` you can use client side to distinguish between different kinds of errors, .e.g:
```json
{
  "code": "CLOUDWATCH_LIMIT_EXCEEDED",
  "message": "Cloud Watch Logs max concurrent queries limit [10] exceeded"
}
```

```go
type PluginError struct {
  Code string
  Message string
}
```

Or maybe something like this if you can get a hold of the limit value easily without parsing strings (fragile if CloudWatch API change the error message).

```json
{
  "code": "CLOUDWATCH_LIMIT_EXCEEDED",
  "payload": {
    "limit": "10"
  }
}
```

```go
type PluginError struct {
  Code string
  Payload map[string]interface{}
}
```

Reply:
Hmm true yeah. In that case this will be close to the original proposal with single generic error with addition of having a significant code string instead of just parsing the message, is that right? I agree that having some code string makes sense. They don't have to be globally unique as they probably need to be understood just between data source backend and frontend.

Reply:
Yeah, sorry for that. 

> proposal with single generic error with addition of having a significant code string instead of just parsing the message, is that right?

Yeah. The original suggestion had HTTP status code part of the solution which might not be needed if there's a code string and always return Bad Request/400. 

> They don't have to be globally unique as they probably need to be understood just between data source backend and frontend.

Yeah exactly.

Reply:
I changed it but in that case the https://github.com/grafana/grafana/pull/39290#discussion_r714198012 comment did not make sense to me. I don't see a way to provide everything from the error to `response.Error()` without changing it quite a bit which I am not sure if makes sense at this point as it is used in lots of places. Let me know what you think not sure how big an issue it is to have this handling in response.go

Reply:
> not sure how big an issue it is to have this handling in response.go

It's not an issue but a code smell handling plugin specific logic in the general response handling code. For now I can live with it though.